sklearn/calibration.py

```diff
@@ -25,6 +27,357 @@
+
+    References
+    ----------
+    .. [1] Receiver-operating characteristic (ROC) plots: a fundamental
+           evaluation tool in clinical medicine, MH Zweig, G Campbell -
+           Clinical chemistry, 1993
+
+    """
+    def __init__(self, base_estimator, strategy='roc', method=None, beta=None,
+                 threshold=None, pos_label=1, cv=3):
+        self.base_estimator = base_estimator
+        self.strategy = strategy
+        self.method = method
+        self.beta = beta
+        self.threshold = threshold
+        self.pos_label = pos_label
+        self.cv = cv
+
+    def fit(self, X, y):
+        """Fit model
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Training data
+
+        y : array-like, shape (n_samples,)
+            Target values. There must be two 2 distinct values
+
+        Returns
+        -------
+        self : object
+            Instance of self
+        """
+        if (not hasattr(self.base_estimator, 'decision_function') and
+                not hasattr(self.base_estimator, 'predict_proba')):
+            raise TypeError('The base_estimator needs to implement either a '
+                            'decision_function or a predict_proba method')
+
+        if self.strategy not in ('roc', 'f_beta', 'max_tpr', 'max_tnr'):
+            raise ValueError('strategy can either be "roc" or "max_tpr" or '
+                             '"max_tnr. Got {} instead'.format(self.strategy))
+
+        if self.method not in (None, 'decision_function', 'predict_proba'):
+            raise ValueError('method param can either be "decision_function" '
+                             'or "predict_proba" or None. '
+                             'Got {} instead'.format(self.method))
+
+        if self.strategy == 'max_tpr' or self.strategy == 'max_tnr':
+            if (not self.threshold or not
+                    isinstance(self.threshold, (int, float))
+                    or not self.threshold >= 0 or not self.threshold <= 1):
+                raise ValueError('parameter threshold must be a number in'
+                                 '[0, 1]. '
+                                 'Got {} instead'.format(self.threshold))
+
+        if self.strategy == 'f_beta':
+            if not self.beta or not isinstance(self.beta, (int, float)):
+                raise ValueError('parameter beta must be a real number.'
+                                 'Got {} instead'.format(type(self.beta)))
+
+        X, y = check_X_y(X, y)
+
+        y_type = type_of_target(y)
+        if y_type != 'binary':
+            raise ValueError('Expected target of binary type. Got {}'.format(
+                y_type))
+
+        self.label_encoder_ = LabelEncoder().fit(y)
+
+        y = self.label_encoder_.transform(y)
+        self.pos_label = self.label_encoder_.transform([self.pos_label])[0]
+
+        if self.cv == 'prefit':
+            self.decision_threshold_ = _CutoffClassifier(
```

Comment:
Nitpick: The code might be simplified by constructing the _CutoffClassifier before the if, then using clone before fitting (indeed, I think this logic of handling `cv` could then be shared with CalibratedClassifierCV)

Reply:
I don't see the benefit in creating the _CutoffClassifier object before the if because it will only be used in the pre-fit case; that's  because the _CutoffClassifier expects a trained base estimator, so in the other case the object will have to be created again. Am I misinterpreting your suggestion?